### ISSUE TITLE
Edit Profile page appears when user log into moodle for the first time

### ISSUE THREAD

**Issue by AmarJaved** [microsoft/o365-moodle#1123]
I have created a new moodle server and installed moodle 3.7 and setup OpenID Connect authentication

When a user logs into moodle for the first time the edit profile page appears, the user cannot navigate anywhere else until the user update the profile. 

I checked the required fields and they are all filled in automatically, when the user logs in for the first time and see the edit profile page, they don't need to make any amendments as all the fields as prepopulated.

While testing I noticed that if I log in for the first time and close the tab (not browser) without clicking on update, the next time I log into moodle in the same browser I am directed to the dashboard (which is the intended behaviour).

I have tried various settings and cannot figure out why the edit profile page appears.

I also have a live moodle server on an moodle 3.1.1 which doesn't have this issue, both server are connecting to the same Azure, the only difference between the 2 is that the live server is using system API whereas the new server is using application API.

The issue with the edit profile page is that I am wanting to disable the permission to allow user to edit their profile, if I disable this permission the user will see an error message when they log in for the first.

Any advice are appreciated
Thanks

**Comment by jamesmcq** [comment]
Hi @AmarJaved - This usually happens when there is a required field the user has to complete, though you mentioned you don't see any. You mentioned the user doesn't have to change anything - so as long as they click "Save changes" they can move on? And if they don't click "Save changes" Moodle won't allow them to visit any other pages?

**Comment by AmarJaved** [comment]
@jamesmcq - that is correct. My first thought was that a required field was missing but they are all filled in.

Also I noticed some odd behaviour where if I close the tab (without clicking on "Save changes") and not the browser and navigate to moodle again, I get directed to the dashboard and can use moodle as a normal user. I won't get asked to save changes again.

However if i close the browser and navigate to moodle in a new browser, I am asked to save changes.

**Comment by byzombies** [comment]
@jamesmcq Just commenting here to confirm the same behavior described by @AmarJaved .

If you disable the editownprofile capability for authenticated users then try to login for the first time as a new user the login itself is successful and the account is created, however the plugin seems to force the user to their profile the first time.

They cannot progress beyond that point, there is only a continue button on the error page, which just puts them in a loop of viewing their profile and getting bounced to the error page.

Logging out and back in makes everything work as expected at this stage.

Currently using the 2017111302 release.

**Comment by gormster** [comment]
I've figured out the issue here, but not sure the best solution.

When a user logs in for the first time, their record in the users table is empty. That's where `get_userinfo` comes in handy – it's called by `authenticate_user_login`  gets the basic user information and populates the `$user` object that is eventually passed to `complete_user_login`, which sets the global `$USER` to that object. However, if you have **local_o365** enabled, `get_userinfo` deliberately returns an empty array. Why?

```php
// auth/oidc/classes/loginflow/base.php:103

// O365 provides custom field mapping, skip OIDC mapping if O365 is present.
$o365installed = $DB->get_record('config_plugins', ['plugin' => 'local_o365', 'name' => 'version']);
if (!empty($o365installed)) {
    return [];
}
```

Of course, the custom field mapping! That happens when **auth_oidc** fires a **user_loggedin** event. When does that fire? Why, in `user_authenticated_hook` of course! And `user_authenticated_hook` gets the user object passed by reference, so it can modify and even totally replace it to its heart's content!

But, of course, it doesn't. It fires an event that updates the user record in the database, but doesn't itself update the passed-in user object. This is in keeping with the design of Moodle's Event system:

> Event observers can not modify event data or interrupt the dispatching of events, it is a one way communication channel.
https://docs.moodle.org/dev/Event_2#What_are_events.3F

Right, so event observers aren't supposed to modify event data. But we know one of our event observers is modifying our user record in a way that we're interested in.

So we have a few choices:

1. Say that a reference to an object isn't really data - it's like a const pointer in C, the pointer itself hasn't changed, just the data it was pointing to - so allowing our event observer to modify/replace it is totally fine; OR
2. Knowing that it's very likely one of our event observers modified the user record, reload it from the database after triggering the event; OR
3. Having determined that **local_o365** is going to provide its own user mapping, actually call that user mapping in `get_userinfo` rather than just returning an empty array and hoping **local_o365** will do its job at some later point.

**Comment by weilai-irl** [comment]
Hi all,

The fix to the issue provided by @gormster in pull request #1197 has been included in the latest release for Moodle 3.7, 3.8 and 3.9 (release 3.7.0.7, 3.8.0.3 and 3.9.0 respectively).

The issue will be closed now.

Regards,
Lai
